res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1136 @@
+                        }.bind(this), true);
+                    } else {
+                        engine.setValue(this.group, "sync_enabled", 0);
+                    }
+                } else {
+                    if (this.longPressTimer !== 0) {
+                        engine.stopTimer(this.longPressTimer);
+                        this.longPressTimer = 0;
+                    }
+                }
+            };
+        }
+        shift() {
+            this.inKey = "quantize";
+            this.type = Button.prototype.types.toggle;
+            this.input = Button.prototype.input;
+        }
+    };
+
+    var LoopToggleButton = class extends Button {
+        constructor(options) {
+            options.inKey = "reloop_exit";
+            options.outKey = "loop_enabled";
+            super(options);
+        }
+        inValueScale() {
+            return 1;
+        }
+    };
+
+    var HotcueButton = class extends Button {
+        constructor(options) {
+            if (options.number === undefined) {
+                throw "No hotcue number specified for HotcueButton constructor";
+            }
+            if (options.colorMapper !== undefined || options.sendRGB !== undefined) {
+                options.colorKey = "hotcue_" + options.number + "_color";
+            }
+            options.outKey = "hotcue_" + options.number + "_enabled";
+            super(options);
+        }
+        unshift() {
+            this.inKey = "hotcue_" + this.number + "_activate";
+        }
+        shift() {
+            this.inKey = "hotcue_" + this.number + "_clear";
+        }
+        output(value) {
+            var outval = this.outValueScale(value);
+            // NOTE: outputColor only handles hotcueColors
+            // and there is no hotcueColor for turning the LED
+            // off. So the `send()` function is responsible for turning the
+            // actual LED off.
+            if (this.colorKey !== undefined && outval !== this.off) {
+                this.outputColor(engine.getValue(this.group, this.colorKey));
+            } else {
+                this.send(outval);
+            }
+        }
+        outputColor(colorCode) {
+            // Sends the color from the colorCode to the controller. This
+            // method will not be called if no colorKey has been specified.
+            if (colorCode === undefined || colorCode < 0 || colorCode > 0xFFFFFF) {
+                print("Ignoring invalid color code '" + colorCode + "' in outputColor()");
+                return;
+            }
+
+            if (this.colorMapper !== undefined) {
+                // This HotcueButton holds a reference to a ColorMapper. This means
+                // that the controller only supports a fixed set of colors, so we
+                // get the MIDI value for the nearest supported color and send it.
+                var nearestColorValue = this.colorMapper.getValueForNearestColor(colorCode);
```

Comment:
we should use `let` or `const` everywhere. `var`'s scoping rules are really weird: https://www.freecodecamp.org/news/var-let-and-const-whats-the-difference/